File: server/utils/audit.ts
import { prisma } from './db'

export async function logAudit(input: {
  entityType: string
  entityId: string
  action: string
  actor: string
  appId?: string | null
  before?: any
  after?: any
}) {
  await prisma.auditLog.create({
    data: {
      entityType: input.entityType,
      entityId: input.entityId,
      action: input.action,
      actor: input.actor,
      appId: input.appId || null,
      before: input.before,
      after: input.after
    }
  })
}

File: backend/app/services/audit.py
import uuid
from sqlalchemy.orm import Session

from app.models import AuditLog


def cuid_like() -> str:
    return uuid.uuid4().hex[:24]


def log_audit(
    db: Session,
    *,
    entity_type: str,
    entity_id: str,
    action: str,
    actor: str,
    app_id: str | None = None,
    before=None,
    after=None,
):
    record = AuditLog(
        id=cuid_like(),
        appId=app_id,
        entityType=entity_type,
        entityId=entity_id,
        action=action,
        actor=actor,
        before=before,
        after=after,
    )
    db.add(record)
    db.commit()
